### PR TITLE
support study-level matrix route when subject is unspecified

### PR DESCRIPTION
**Pull request by JackVCurtis**
Screenshot 2024-03-15 at 9 57 39 AM

Rather than adjust anything on the backend I've just refactored the GraphPage to query the study for all participants if the value for the participant is missing, and then query the data endpoint once for each participant. It makes the new `Graph` component stateful, which we usually avoid for non-pages, but simplifies parallelization relative to the old `co(*function` approach on the backend

I also removed a bunch of unused code related to handling scroll/zoom with JS, since the new CSS handles that for us.

### CODE REVIEW

File: views/components/Matrix.d3.js [hunk 1]
```diff
@@ -370,7 +252,9 @@ export default class Matrix {
       } else {
         xAxisForDatesData.push({ day, marker: 'N/A' })
       }
-      startDate.setDate(startDate.getDate() + 1)
+      if (startDate) {
+        startDate.setDate(startDate.getDate() + 1)
+      }
```
Comment (Elkrival) [inline views/components/Matrix.d3.js:258]:
```suggestion
      if (startDate)  startDate.setDate(startDate.getDate() + 1)
      
```

File: views/components/Matrix.d3.js [hunk 2]
```diff
@@ -357,7 +239,7 @@ export default class Matrix {
     for (let i = firstDay; i < this.lastDayForFilter; i++) {
       const day = i + 1
 
-      if (startDate && startDate.getDay() == 0 || startDate.getDay() == 6) {
+      if (startDate && (startDate.getDay() == 0 || startDate.getDay() == 6)) {
```
Comment (Elkrival) [inline views/components/Matrix.d3.js:243]:
```suggestion
      if (startDate && (startDate.getDay() === 0 || startDate.getDay() === 6)) {
```

How did they come up with 6..?

File: views/components/Matrix.d3.js [hunk 3]
```diff
@@ -250,98 +224,6 @@ export default class Matrix {
       .style('fill', DARK_GREY)
   }
 
-  scaleAxes = (transform) => {
```
Comment (Elkrival) [inline views/components/Matrix.d3.js:253]: 🎉

File: views/pages/GraphPage.jsx [hunk 4]
```diff
@@ -1,60 +1,37 @@
...
-    maxDay: 1,
-  })
-  const [openStat, setOpenStat] = React.useState(false)
-  const downloadPng = () =>
-    canvasRef.current.toBlob((blob) => {
-      FileSaver.saveAs(blob, `${subject}.png`)
-    })
+  const [participants, setParticipants] = useState([])
+  useEffect(() => {
+    const asyncEffect = async () => {
```
Comment (Elkrival) [inline views/pages/GraphPage.jsx]: what do you think about moving this function out of the use effect and naming it something like fetchParticipants..?

File: views/components/Matrix.d3.js [hunk 5]
```diff
@@ -60,7 +57,7 @@ export default class Matrix {
       .select(this.el)
       .append('svg')
       .attr('width', () => '100%')
-      .attr('height', () => '100%')
+      .attr('height', () => data.length * this.cardSize + margin.top + margin.bottom)
```
Comment (Elkrival) [inline views/components/Matrix.d3.js]: I'm curious how this number came about, also what do you think about moving it to a constant.
```suggestion
      const calculatedHeight = data.length * this.cardSize + margin.top + margin.bottom
      
      
      .attr('height', () => calculatedHeight)
```

File: test/fixtures.js [hunk 6]
```diff
@@ -308,3 +308,66 @@ export const createAssessmentDayData = (overrides = {}) => ({
...
+                    "#41ab5d",
+                    "#238b45",
+                    "#006d2c",
+                    "#00441b"
+                ],
+                "category": "phone_use",
+                "analysis": "phone_power_activityScores_hourly"
+            },
+        ],
+        "consentDate": "2022-06-02T00:00:00.000Z"
```
Comment (Elkrival) [inline test/fixtures.js:369]: should we add a test for the missing consentDate? 
Or maybe take it out entirely and use the daysInStudy attribute from the same collection. 

Thoughts..?

**Review by mikestone14 — COMMENTED** [review]
overall this looks good, I'm just wondering how many participants might need to get loaded on this page. If we're making an API call per participant could we be making lots of API calls to get this to load.

**Review by Elkrival — APPROVED** [review]
This is looking good, some adjustments to naming variables and i'm curious about a few things.

File: views/components/Matrix.d3.js (hunk 2 above)
Comment (JackVCurtis) [reply views/components/Matrix.d3.js:243]: 6 is the numeric value for Saturday, 0 is for Sunday

File: views/pages/GraphPage.jsx (hunk 4 above)
Comment (mikestone14) [reply views/pages/GraphPage.jsx]: this is a common pattern for functions that are only called inside the useEffect. If this was moved outside the useEffect then the useEffect should technically depend on it, which would in turn mean using something like a useCallback for this function but it's just easier keeping it here 👍

**Comment by JackVCurtis** [comment]
> overall this looks good, I'm just wondering how many participants might need to get loaded on this page. If we're making an API call per participant could we be making lots of API calls to get this to load.

Yeah, definitely it could involve a lot of calls. The problem is there isn't really any way to get this data except a per-subject query. The old version of this feature used parallelism on the backend with the `co` package, but that's IMO an even less performant approach because we're doing the same queries but we have to keep all the results in memory at once, plus the overhead of managing the coroutines. Node.js isn't really built for backend parallelism, I figured parallelizing on the client side instead would be both simpler and more practical given Node's single-threaded-async model. Node handles sequential web requests really well, and with this approach once each request is resolved the data can be cleaned out of memory.

File: views/pages/GraphPage.jsx (hunk 4 above)
Comment (Elkrival) [reply views/pages/GraphPage.jsx]: Thanks for sharing that @mikestone14

File: views/components/Matrix.d3.js (hunk 5 above)
Comment (JackVCurtis) [reply views/components/Matrix.d3.js]: Non-fixed heights were behaving weirdly when I added multiple matrices, so I calculated by the cell height.

**Pull request merged**